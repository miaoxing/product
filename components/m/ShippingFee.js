import $ from 'miaoxing';
import {useEffect, useState} from 'react';
import {Text} from '@fower/taro';
import PropTypes from 'prop-types';

const ShippingFee = ({productId}) => {
  const [shippingTpl, setShippingTpl] = useState();

  useEffect(() => {
    $.http({
      url: $.apiUrl('products/%s/shipping-tpl?filterRulesByCity=1', productId),
    }).then(({ret}) => {
      if (ret.isErr()) {
        return;
      }
      setShippingTpl(ret);
    });
  }, [productId]);

  if (!shippingTpl) {
    return <Text/>;
  }

  let fee = '';
  if (shippingTpl.data.isFreeShipping) {
    fee = '包邮';
  } else {
    const rule = shippingTpl.data.rules[0];
    fee = rule.service.name + '：￥' + rule.startFee;
  }

  return (
    <Text>
      {shippingTpl.city && `到${shippingTpl.city}`}
      {' '}
      {fee}
    </Text>
  );
};

ShippingFee.propTypes = {
  productId: PropTypes.number.isRequired,
};

export default ShippingFee;
