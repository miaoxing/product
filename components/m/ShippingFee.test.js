import ShippingFee from './ShippingFee';
import {render, waitFor} from '@testing-library/react';
import $, {Ret} from 'miaoxing';
import {createPromise, bootstrap} from '@mxjs/test';

bootstrap();

describe('ShippingFee', () => {
  test('basic', async () => {
    const promise = createPromise();

    $.http = jest.fn()
      .mockImplementationOnce(() => promise.resolve({
        ret: Ret.suc({
          data: {
            rules: [
              {
                startFee: '9.8',
                service: {
                  name: '快递',
                },
              },
            ],
          },
          city: '深圳',
        }),
      }));

    const {container} = render(<ShippingFee productId={1}/>);

    await waitFor(() => {
      expect(container.textContent).toContain('9.8');
    });

    expect(container).toMatchSnapshot();
    expect($.http).toMatchSnapshot();
  });

  test('without city', async () => {
    const promise = createPromise();

    $.http = jest.fn()
      .mockImplementationOnce(() => promise.resolve({
        ret: Ret.suc({
          data: {
            rules: [
              {
                startFee: '9.8',
                service: {
                  name: '快递',
                },
              },
            ],
          },
          city: '',
        }),
      }));

    const {container} = render(<ShippingFee productId={1}/>);

    await waitFor(() => {
      expect(container.textContent).toContain('9.8');
    });

    expect(container).toMatchSnapshot();
    expect($.http).toMatchSnapshot();
  });

  test('free shipping', async () => {
    const promise = createPromise();

    $.http = jest.fn()
      .mockImplementationOnce(() => promise.resolve({
        ret: Ret.suc({
          data: {
            isFreeShipping: true,
            rules: [],
          },
          city: '广州',
        }),
      }));

    const {container} = render(<ShippingFee productId={1}/>);

    await waitFor(() => {
      expect(container.textContent).toContain('包邮');
    });

    expect(container).toMatchSnapshot();
    expect($.http).toMatchSnapshot();
  });

  test('start fee 0', async () => {
    const promise = createPromise();

    $.http = jest.fn()
      .mockImplementationOnce(() => promise.resolve({
        ret: Ret.suc({
          data: {
            isFreeShipping: false,
            rules: [
              {
                startFee: 0,
                service: {
                  name: '快递',
                },
              },
            ],
          },
          city: '北京',
        }),
      }));

    const {container} = render(<ShippingFee productId={1}/>);

    await waitFor(() => {
      expect(container.textContent).toContain('￥0');
    });

    expect(container).toMatchSnapshot();
    expect($.http).toMatchSnapshot();
  });
});
