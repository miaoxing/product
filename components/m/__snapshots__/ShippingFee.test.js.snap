// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShippingFee basic 1`] = `
<div>
  <taro-text-core>
    到深圳
     
    快递：￥9.8
  </taro-text-core>
</div>
`;

exports[`ShippingFee basic 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "url": "/api/products/1/shipping-tpl?filterRulesByCity=1",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;

exports[`ShippingFee free shipping 1`] = `
<div>
  <taro-text-core>
    到广州
     
    包邮
  </taro-text-core>
</div>
`;

exports[`ShippingFee free shipping 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "url": "/api/products/1/shipping-tpl?filterRulesByCity=1",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;

exports[`ShippingFee start fee 0 1`] = `
<div>
  <taro-text-core>
    到北京
     
    快递：￥0
  </taro-text-core>
</div>
`;

exports[`ShippingFee start fee 0 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "url": "/api/products/1/shipping-tpl?filterRulesByCity=1",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;

exports[`ShippingFee without city 1`] = `
<div>
  <taro-text-core>
    
     
    快递：￥9.8
  </taro-text-core>
</div>
`;

exports[`ShippingFee without city 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "url": "/api/products/1/shipping-tpl?filterRulesByCity=1",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;
